pkg/mutation/schema/node.go

```diff
@@ -0,0 +1,180 @@
+package schema
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/path/parser"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/types"
+)
+
+type references map[types.ID]bool
+
+func (c references) String() string {
+	var keys []string
+	for k := range c {
+		keys = append(keys, fmt.Sprintf("%q", k.String()))
+	}
+	sort.Strings(keys)
+	return fmt.Sprintf("[%s]", strings.Join(keys, ","))
+}
+
+// node is an element of an implicit schema.
+// Allows for the definition of overlapping schemas. See Add.
+type node struct {
+	// references tracks the Mutations which reference this part of the schema tree.
+	references references
+
+	// children is the set of child Nodes a this location in the schema.
+	// Each node defines a distinct child definition. If multiple Nodes are defined
+	// for the same child, then there is a schema conflict.
+	children map[string]map[parser.NodeType]node
+}
+
+// Add inserts the provided path, linked to the given ID.
+//
+// Returns the set of conflicts detected while adding the path. Conflicts occur
+// when elements with the same path have different types, for example:
+//
+// spec.containers[name: foo].image
+// spec.containers.image
+//
+// If the returned references is non-nil it contains at least two elements, one
+// of which is the passed id.
+func (n *node) Add(id types.ID, path []parser.Node) references {
+	if n.references == nil {
+		n.references = make(map[types.ID]bool)
+	}
+	// This node is referenced by the passed ID.
+	n.references[id] = true
+
+	// Base case; either there is no more path to validate or this must be an
+	// object or a base type.
+	if len(path) == 0 {
+		return nil
+	}
+
+	// Initialize child within n.children.
+	childNode := path[0]
+	if n.children == nil {
+		n.children = make(map[string]map[parser.NodeType]node)
+	}
+	childKey := childNode.Key()
```

Comment:
We actually don't want to store a list by its key field... the key field is really part of it's schema.

if I see

`spec.containers[name: "foo"]`,

that's telling me `spec.containers` is a list with a unique key field of "name".

If I see another `location` that says `spec.containers[image; "my/docker:image"]

that's telling me `spec.containers` is a list with a unique key field of "image".

Since the unique key field must always be the same, these two schemas are in conflict.

TL;DR the key field of a list is used for conflict detection, not pointing to a child value.

Currently this PR isn't detecting unique key conflicts, which it should.

Reply:
Okay, so you are saying these two schemas are in conflict:

```
spec.containers[name: "foo"]
spec.containers[image: "bar"]
```

Reply:
Done.